tools/rabbitmq-consumer/src/RabbitMQConsumer/Lib.hs

```diff
@@ -0,0 +1,232 @@
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+{-# LANGUAGE OverloadedStrings #-}
+
+module RabbitMQConsumer.Lib where
+
+import Data.Aeson
+import Data.Aeson.Encode.Pretty
+import Data.ByteString.Lazy.Char8 qualified as BL
+import Data.Domain (Domain)
+import Data.Text.Lazy.Encoding qualified as TL
+import Imports
+import Network.AMQP
+import Network.Socket
+import Options.Applicative
+import Wire.API.Federation.BackendNotifications (BackendNotification (..))
+import Wire.API.MakesFederatedCall (Component)
+
+main :: IO ()
+main = do
+  opts <- execParser (info (helper <*> optsParser) desc)
+  conn <- openConnection' opts.host opts.port opts.vhost opts.username opts.password
+  chan <- openChannel conn
+  qos chan 0 1 False
+  done <- newEmptyMVar
+  case opts.cmd of
+    Interactive -> void $ consumeMsgs chan opts.queue Ack (interactive done opts)
+    Head -> do
+      runTimerAsync done opts.timeoutSec
+      void $ consumeMsgs chan opts.queue Ack (printHead done opts)
+    DropHead dhOpts -> do
+      runTimerAsync done opts.timeoutSec
+      void $ consumeMsgs chan opts.queue Ack (dropHead done opts dhOpts)
+  takeMVar done
+  closeConnection conn
+  putStrLn "connection closed"
+  where
+    desc = header "rabbitmq-consumer" <> progDesc "CLI tool to consume messages from a RabbitMQ queue" <> fullDesc
+
+    printHead :: MVar () -> Opts -> (Message, Envelope) -> IO ()
+    printHead done opts (msg, _env) = do
+      putStrLn $ displayMessage opts msg
+      void $ tryPutMVar done ()
+
+    dropHead :: MVar () -> Opts -> DropHeadOpts -> (Message, Envelope) -> IO ()
+    dropHead done opts dhOpts (msg, env) = do
+      putStrLn $ displayMessage opts msg
+      case decode @BackendNotification msg.msgBody of
+        Nothing -> putStrLn "failed to decode message body"
+        Just bn -> do
+          if bn.path == dhOpts.path
+            then do
+              putStrLn "dropping message"
+              nackEnv env
+            else do
+              putStrLn "path does not match. keeping message"
+      void $ tryPutMVar done ()
+
+    interactive :: MVar () -> Opts -> (Message, Envelope) -> IO ()
+    interactive done opts (msg, env) = do
+      putStrLn $ displayMessage opts msg
+      putStrLn $ "type 'drop' to drop the message and terminate, or press enter to terminate without dropping the message"
+      input <- getLine
+      if input == "drop"
+        then do
+          ackEnv env
+          putStrLn "message dropped"
+        else putStrLn "message not dropped"
+      void $ tryPutMVar done ()
+
+    displayMessage :: Opts -> Message -> String
+    displayMessage opts msg =
+      intercalate
+        "\n"
+        [ "vhost: " <> cs opts.vhost,
+          "queue: " <> cs opts.queue,
+          "timestamp: " <> show msg.msgTimestamp,
+          "received message: \n" <> BL.unpack (maybe msg.msgBody encodePretty (decode @BackendNotification' msg.msgBody))
```

Comment:
We should handle the decode failure here and print the JSON as is with a warning that it failed to parse and the parse error. We can use `eitherDecode` to get the parse error.

Reply:
Sorry needed more coffee :coffee: